Create a tutorial for using moveit_opw_kinematics_plugin (#607)
ros-planning/moveit_tutorials: doc/opw_kinematics/opw_kinematics_tutorial.rst

joint-zero offsets. Given this structure, no other setup is required.

Installation
-------------

The `opw_kinematics_plugin https://github.com/JeroenDM/moveit_opw_kinematics_plugin`_ can be installed from packages: ::

  sudo apt install ros-noetic-opw-kinematics-plugin

Usage
------

- Find the MoveIt `kinematics.yaml <../kinematics_configuration/kinematics_configuration_tutorial.html>`_ file created for your robot.
- Replace ``kinematics_solver: kdl_kinematics_plugin/KDLKinematicsPlugin`` (or similar) with ``kinematics_solver: moveit_opw_kinematics_plugin/MoveItOPWKinematicsPlugin``
- Set parameters to describe the geometry of your manipulator. The following is an example for the KUKA KR 6 R700: ::

  manipulator:
    kinematics_solver: moveit_opw_kinematics_plugin/MoveItOPWKinematicsPlugin
    opw_kinematics_geometric_parameters:
      a1:  0.025
      a2: -0.035
      b:   0.000
      c1:  0.400
      c2:  0.315
      c3:  0.365
      c4:  0.080
    opw_kinematics_joint_offsets: [0.0, -1.57079632679, 0, 0, 0, 0]
    opw_kinematics_joint_sign_corrections: [-1, 1, 1, -1, 1, -1]

> The meaning of the parameters can best be understood with a sketch and some tinkering. The plugin will print ``ROS_ERROR``s on startup if they

Reviewer: Not sure the `s` works there after the `ROS_ERROR`.

At least GH highlighter seems to get confused.

Author: that seems easily confused, but http://rst.ninjs.org/ found some more issues, let's see if that also fixes travis

Based on the paper *An Analytical Solution of the Inverse Kinematics Problem of Industrial Serial Manipulators
with an Ortho-parallel Basis and a Spherical Wrist* by Mathias Brandstötter, Arthur Angerer, and Michael Hofbaur.

Purpose
-------

This package is meant to provide a simpler alternative to IK-Fast based solutions in situations where one has an
industrial robot with a parallel base and spherical wrist. This configuration is extremely common in industrial robots.

The kinematics are parameterized by 7 primary values taken directly from the robot's spec sheet and a set of
joint-zero offsets. Given this structure, no other setup is required.

Installation
-------------

The `opw_kinematics_plugin https://github.com/JeroenDM/moveit_opw_kinematics_plugin`_ can be installed from packages: ::

  sudo apt install ros-noetic-opw-kinematics-plugin

Usage
------

- Find the MoveIt `kinematics.yaml <../kinematics_configuration/kinematics_configuration_tutorial.html>`_ file created for your robot.
- Replace ``kinematics_solver: kdl_kinematics_plugin/KDLKinematicsPlugin`` (or similar) with ``kinematics_solver: moveit_opw_kinematics_plugin/MoveItOPWKinematicsPlugin``

> - Set parameters to describe the geometry of your manipulator. The following is an example for Kuka Kr6R700: ::

Reviewer: ```suggestion
- Set parameters to describe the geometry of your manipulator. The following is an example for the KUKA KR6 R700: ::
```

Author: just googled, it should actually be KR 6 R700
```suggestion
- Set parameters to describe the geometry of your manipulator. The following is an example for the KUKA KR 6 R700: ::
```

  manipulator:
    kinematics_solver: moveit_opw_kinematics_plugin/MoveItOPWKinematicsPlugin
    opw_kinematics_geometric_parameters:
      a1:  0.025
      a2: -0.035
      b:   0.000
      c1:  0.400
      c2:  0.315
      c3:  0.365
      c4:  0.080
    opw_kinematics_joint_offsets: [0.0, -1.57079632679, 0, 0, 0, 0]
    opw_kinematics_joint_sign_corrections: [-1, 1, 1, -1, 1, -1]

> The meaning of the parameters can best be understood with a sketch and some tinkering. The plugin will throw ROS_ERRORs on startup if they

Reviewer: ```suggestion
The meaning of the parameters can best be understood with a sketch and some tinkering. The plugin will print ``ROS_ERROR``s on startup if they
```

do not match your URDF, so you can safely guess and test if needed:

.. image:: images/opw.png
    
Sharing OPW descriptions
-------------------------

> We plan to collect OPW parameter sets as part of the ROS-I robot description packages. This has already started for Fanuc.

Reviewer: ```suggestion
We plan to collect OPW parameter sets as part of the ROS-Industrial robot support packages. This has already started for FANUC.
```

not sure this is really specific to support packages though.

Author: I'm not sure I understand your what you are not sure about. The intention of this paragraph was to include somewhere that collaboration on parameter files is desired

> Taking the `LR Mate 200iB <https://github.com/ros-industrial/fanuc/blob/kinetic-devel/fanuc_lrmate200ib_support/config/opw_parameters_lrmate200ib.yaml>`_ as

Reviewer: ```suggestion
Taking the `LR Mate 200iB <https://github.com/ros-industrial/fanuc/blob/3ea2842baca3184cc621071b785cbf0c588a4046/fanuc_lrmate200ib_support/config/opw_parameters_lrmate200ib.yaml>`_ as
```

an example you can reduce your *kinematics.yaml* to the following header: ::

  manipulator:
    kinematics_solver: moveit_opw_kinematics_plugin/MoveItOPWKinematicsPlugin

> and then add second line of the following to your *launch/planning_context.launch*: ::

Reviewer: ```suggestion
and then add a ``rosparam`` ``load`` line to your *launch/planning_context.launch* which causes the parameters in that file to be loaded onto the parameter server: ::
```

  <!-- Load default settings for kinematics; these settings are overridden by settings in a node's namespace -->
  <group ns="$(arg robot_description)_kinematics">
    <rosparam command="load" file="$(find opw_tutorial)/config/kinematics.yaml"/>
    <rosparam command="load" ns="manipulator" file="$(find fanuc_lrmate200ib_support)/config/opw_parameters_lrmate200ib.yaml"/>
  </group>

Note that the *ns* parameter has to match the name you gave your planning group during the setup.

> This can also be done directly within the MoveIt setup assistant. On the *Define Planning Group* page, select the *opw_parameters*

Reviewer: ```suggestion
The MoveIt Setup Assistant can automatically insert that line, removing the need to manually edit the ``planning_context.launch`` file. On the *Define Planning Group* page, select the *opw_parameters*
```

OPW Kinematics Solver for Industrial Manipulators
==================================================

Intro
------

Taken from `opw_kinematics <https://github.com/Jmeyer1292/opw_kinematics>`_: 
A simple, analytical inverse kinematic library for industrial robots with parallel bases and spherical wrists.
Based on the paper *An Analytical Solution of the Inverse Kinematics Problem of Industrial Serial Manipulators
with an Ortho-parallel Basis and a Spherical Wrist* by Mathias Brandstötter, Arthur Angerer, and Michael Hofbaur.

Purpose
-------

This package is meant to provide a simpler alternative to IK-Fast based solutions in situations where one has an
industrial robot with a parallel base and spherical wrist. This configuration is extremely common in industrial robots.

The kinematics are parameterized by 7 primary values taken directly from the robot's spec sheet and a set of
joint-zero offsets. Given this structure, no other setup is required.

Installation
-------------

> The `opw_kinematics_plugin <https://github.com/JeroenDM/moveit_opw_kinematics_plugin>`_ can be installed from packages: ::

Reviewer: ```suggestion
The `opw_kinematics_plugin <https://github.com/JeroenDM/moveit_opw_kinematics_plugin>`_ can be installed using ``apt`` on Ubuntu and Debian: ::
```

  sudo apt install ros-noetic-opw-kinematics-plugin

Usage
------

- Find the MoveIt `kinematics.yaml <../kinematics_configuration/kinematics_configuration_tutorial.html>`_ file created for your robot.

> - Replace ``kinematics_solver: kdl_kinematics_plugin/KDLKinematicsPlugin`` (or similar) with ``kinematics_solver: moveit_opw_kinematics_plugin/MoveItOPWKinematicsPlugin``

Reviewer: ```suggestion
- Replace ``kinematics_solver: kdl_kinematics_plugin/KDLKinematicsPlugin`` (or similar) with ``kinematics_solver: moveit_opw_kinematics_plugin/MoveItOPWKinematicsPlugin``.
```

- Set parameters to describe the geometry of your manipulator.

> The following is an example for the KUKA KR 6 R700: ::

Reviewer: ```suggestion
The following is an example of the parameters needed for the KUKA KR 6 R700: ::
```

>   manipulator:
>     kinematics_solver: moveit_opw_kinematics_plugin/MoveItOPWKinematicsPlugin
>     opw_kinematics_geometric_parameters:
>       a1:  0.025
>       a2: -0.035
>       b:   0.000
>       c1:  0.400
>       c2:  0.315
>       c3:  0.365
>       c4:  0.080
>     opw_kinematics_joint_offsets: [0.0, -1.57079632679, 0, 0, 0, 0]
>     opw_kinematics_joint_sign_corrections: [-1, 1, 1, -1, 1, -1]

Reviewer: ```suggestion
.. code-block:: yaml

   manipulator:
     kinematics_solver: moveit_opw_kinematics_plugin/MoveItOPWKinematicsPlugin
     opw_kinematics_geometric_parameters:
       a1:  0.025
       a2: -0.035
       b:   0.000
       c1:  0.400
       c2:  0.315
       c3:  0.365
       c4:  0.080
     opw_kinematics_joint_offsets: [0.0, -1.57079632679, 0, 0, 0, 0]
     opw_kinematics_joint_sign_corrections: [-1, 1, 1, -1, 1, -1]
```

And I believe you don't need the `::` any more with this.

The meaning of the parameters can best be understood with a sketch and some tinkering. The plugin will print a ``ROS_ERROR`` on startup if they
do not match your URDF, so you can safely guess and test if needed:

.. image:: images/opw.png
    
Sharing OPW descriptions
-------------------------

We plan to collect OPW parameter sets as part of the ROS-Industrial robot support packages. This has already started for FANUC.
Taking the `LR Mate 200iB <https://github.com/ros-industrial/fanuc/blob/3ea2842baca3184cc621071b785cbf0c588a4046/fanuc_lrmate200ib_support/config/opw_parameters_lrmate200ib.yaml>`_ as

> an example you can reduce your *kinematics.yaml* to the following header: ::

Reviewer: ```suggestion
an example you can reduce your ``kinematics.yaml`` to the following: ::
```

  manipulator:
    kinematics_solver: moveit_opw_kinematics_plugin/MoveItOPWKinematicsPlugin

> and then add a ``rosparam`` ``load`` line to your *launch/planning_context.launch* which causes the parameters in that file to be loaded onto the parameter server: ::

Reviewer: ```suggestion
and then add a ``rosparam`` ``load`` line to your ``launch/planning_context.launch`` which causes the parameters in that file to be loaded onto the parameter server: ::
```

  <!-- Load default settings for kinematics; these settings are overridden by settings in a node's namespace -->
  <group ns="$(arg robot_description)_kinematics">
    <rosparam command="load" file="$(find opw_tutorial)/config/kinematics.yaml"/>
    <rosparam command="load" ns="manipulator" file="$(find fanuc_lrmate200ib_support)/config/opw_parameters_lrmate200ib.yaml"/>
  </group>

> Note that the *ns* parameter has to match the name you gave your planning group during the setup.

Reviewer: ```suggestion
Note that the ``ns`` parameter has to match the name you gave your planning group during the setup.
```

The MoveIt Setup Assistant can automatically insert that line, removing the need to manually edit the ``planning_context.launch`` file. On the *Define Planning Group* page, select the *opw_parameters*
file for your robot in the *Kin. parameters file* field

> .. image:: images/assistant.png

Reviewer: Would be great to highlight the input field in the screenshot (with a req square or something).

with an Ortho-parallel Basis and a Spherical Wrist* by Mathias Brandstötter, Arthur Angerer, and Michael Hofbaur.

Purpose
-------

This package is meant to provide a simpler alternative to IK-Fast based solutions in situations where one has an
industrial robot with a parallel base and spherical wrist. This configuration is extremely common in industrial robots.

The kinematics are parameterized by 7 primary values taken directly from the robot's spec sheet and a set of
joint-zero offsets. Given this structure, no other setup is required.

Installation
-------------

The `opw_kinematics_plugin <https://github.com/JeroenDM/moveit_opw_kinematics_plugin>`_ can be installed using ``apt`` on Ubuntu and Debian: ::

  sudo apt install ros-noetic-opw-kinematics-plugin

Usage
------

- Find the MoveIt `kinematics.yaml <../kinematics_configuration/kinematics_configuration_tutorial.html>`_ file created for your robot.
- Replace ``kinematics_solver: kdl_kinematics_plugin/KDLKinematicsPlugin`` (or similar) with ``kinematics_solver: moveit_opw_kinematics_plugin/MoveItOPWKinematicsPlugin``.
- Set parameters to describe the geometry of your manipulator.

> The following is an example of the parameters needed for the KUKA KR 6 R700: ::

Author: ```suggestion
The following is an example of the parameters needed for the KUKA KR 6 R700:
```

.. code-block:: yaml

   manipulator:
     kinematics_solver: moveit_opw_kinematics_plugin/MoveItOPWKinematicsPlugin
     opw_kinematics_geometric_parameters:
       a1:  0.025
       a2: -0.035
       b:   0.000
       c1:  0.400
       c2:  0.315
       c3:  0.365
       c4:  0.080
     opw_kinematics_joint_offsets: [0.0, -1.57079632679, 0, 0, 0, 0]
     opw_kinematics_joint_sign_corrections: [-1, 1, 1, -1, 1, -1]

The meaning of the parameters can best be understood with a sketch and some tinkering. The plugin will print a ``ROS_ERROR`` on startup if they
do not match your URDF, so you can safely guess and test if needed:

.. image:: images/opw.png
    
Sharing OPW descriptions
-------------------------

We plan to collect OPW parameter sets as part of the ROS-Industrial robot support packages. This has already started for FANUC.
Taking the `LR Mate 200iB <https://github.com/ros-industrial/fanuc/blob/3ea2842baca3184cc621071b785cbf0c588a4046/fanuc_lrmate200ib_support/config/opw_parameters_lrmate200ib.yaml>`_ as

> an example you can reduce your ``kinematics.yaml`` to the following: ::

Author: ```suggestion
an example you can reduce your ``kinematics.yaml`` to the following:
```

>   manipulator:
>     kinematics_solver: moveit_opw_kinematics_plugin/MoveItOPWKinematicsPlugin

Reviewer: ```suggestion
.. code-block:: yaml

   manipulator:
     kinematics_solver: moveit_opw_kinematics_plugin/MoveItOPWKinematicsPlugin
```

> and then add a ``rosparam`` ``load`` line to your ``launch/planning_context.launch`` which causes the parameters in that file to be loaded onto the parameter server: ::

Author: ```suggestion
and then add a ``rosparam`` ``load`` line to your ``launch/planning_context.launch`` which causes the parameters in that file to be loaded onto the parameter server:
```

>   <!-- Load default settings for kinematics; these settings are overridden by settings in a node's namespace -->
>   <group ns="$(arg robot_description)_kinematics">
>     <rosparam command="load" file="$(find opw_tutorial)/config/kinematics.yaml"/>
>     <rosparam command="load" ns="manipulator" file="$(find fanuc_lrmate200ib_support)/config/opw_parameters_lrmate200ib.yaml"/>
>   </group>

Reviewer: ```suggestion
.. code-block:: xml

   <!-- Load default settings for kinematics; these settings are overridden by settings in a node's namespace -->
   <group ns="$(arg robot_description)_kinematics">
     <rosparam command="load" file="$(find opw_tutorial)/config/kinematics.yaml"/>
     <rosparam command="load" ns="manipulator" file="$(find fanuc_lrmate200ib_support)/config/opw_parameters_lrmate200ib.yaml"/>
   </group>
```

- Replace ``kinematics_solver: kdl_kinematics_plugin/KDLKinematicsPlugin`` (or similar) with ``kinematics_solver: moveit_opw_kinematics_plugin/MoveItOPWKinematicsPlugin``.
- Set parameters to describe the geometry of your manipulator.

The following is an example of the parameters needed for the KUKA KR 6 R700:

.. code-block:: yaml

   manipulator:
     kinematics_solver: moveit_opw_kinematics_plugin/MoveItOPWKinematicsPlugin
     opw_kinematics_geometric_parameters:
       a1:  0.025
       a2: -0.035
       b:   0.000
       c1:  0.400
       c2:  0.315
       c3:  0.365
       c4:  0.080
     opw_kinematics_joint_offsets: [0.0, -1.57079632679, 0, 0, 0, 0]
     opw_kinematics_joint_sign_corrections: [-1, 1, 1, -1, 1, -1]

The meaning of the parameters can best be understood with a sketch and some tinkering. The plugin will print a ``ROS_ERROR`` on startup if they
do not match your URDF, so you can safely guess and test if needed:

.. image:: images/opw.png
    
Sharing OPW descriptions
-------------------------

We plan to collect OPW parameter sets as part of the ROS-Industrial robot support packages. This has already started for FANUC.

> Taking the `M-10iA <https://github.com/ros-industrial/fanuc/blob/kinetic-devel/fanuc_m10ia_support/config/opw_parameters_m10ia.yaml>`_ as

Reviewer: ```suggestion
Taking the `M-10iA <https://github.com/ros-industrial/fanuc/blob/3ea2842baca3184cc621071b785cbf0c588a4046/fanuc_m10ia_support/config/opw_parameters_m10ia.yaml>`_ as
```
I would probably still recommend to use a permalink.

Author: Oh, I got confused why your previous suggestion wasn't applied any longer but now I see that your original suggestion was from branch -> hash already and I reverted that with my last change. Sorry!

> Taking the `LR Mate 200iB <https://github.com/ros-industrial/fanuc/blob/kinetic-devel/fanuc_m10ia_support/config/opw_parameters_m10ia.yaml>`_ as

Author: ```suggestion
Taking the `M-10iA <https://github.com/ros-industrial/fanuc/blob/kinetic-devel/fanuc_m10ia_support/config/opw_parameters_m10ia.yaml>`_ as
```

> Taking the `LR Mate 200iB <https://github.com/ros-industrial/fanuc/blob/3ea2842baca3184cc621071b785cbf0c588a4046/fanuc_lrmate200ib_support/config/opw_parameters_lrmate200ib.yaml>`_ as

Author: ```suggestion
Taking the `LR Mate 200iB <https://github.com/ros-industrial/fanuc/blob/kinetic-devel/fanuc_m10ia_support/config/opw_parameters_m10ia.yaml>`_ as
```

> Taking the `LR Mate 200iB <https://github.com/ros-industrial/fanuc/blob/3ea2842baca3184cc621071b785cbf0c588a4046/fanuc_lrmate200ib_support/config/opw_parameters_lrmate200ib.yaml>`_ as

Author: ```suggestion
Taking the `M-10iA <https://github.com/ros-industrial/fanuc/blob/3ea2842baca3184cc621071b785cbf0c588a4046/fanuc_lrmate200ib_support/config/opw_parameters_lrmate200ib.yaml>`_ as
```
